src/borg/archiver.py

```diff
@@ -204,18 +204,55 @@ def do_check(self, args, repository):
             if not yes(msg, false_msg="Aborting.", truish=('YES', ),
                        env_var_override='BORG_CHECK_I_KNOW_WHAT_I_AM_DOING'):
                 return EXIT_ERROR
-        if args.repo_only and args.verify_data:
-            self.print_error("--repository-only and --verify-data contradict each other. Please select one.")
+
+        if args.repo_only and any((args.verify_data, args.first, args.last)):
+            self.print_error("--repository-only contradicts --first, --last "
+                             "and --verify-data arguments.")
             return EXIT_ERROR
+
         if not args.archives_only:
             if not repository.check(repair=args.repair, save_space=args.save_space):
                 return EXIT_WARNING
-        if not args.repo_only and not ArchiveChecker().check(
-                repository, repair=args.repair, archive=args.location.archive,
-                last=args.last, prefix=args.prefix, verify_data=args.verify_data,
-                save_space=args.save_space):
-            return EXIT_WARNING
-        return EXIT_SUCCESS
+
+        if not args.repo_only:
+            logger.info('Starting archive consistency check...')
+            with ArchiveChecker(args, repository) as archive_checker:
+                manifest = archive_checker.manifest
+                all_archives_count = len(manifest.archives.list())
+
+                if args.verify_data:
+                    archive_checker.verify_data()
+
+                if args.location.archive:
+                    archive_info = manifest.archives.get(args.location.archive)
+                    if archive_info is None:
+                        logger.error('Archive %s not found.' % args.location.archive)
+                        return EXIT_ERROR
+                    archives = (archive_info,)
+                elif any((args.first, args.last, args.prefix)):
+                    archives = self._get_filtered_archives(args, manifest)
+                else:
+                    archives = manifest.archives.list(prefix=args.prefix)
```

Comment:
that looks strange.
232 will trigger if args.prefix is truish.
but in 235 you give args.prefix as a param (can it be anything else than None or '' here?).